docs/design/proposals/stateless-csi-plugin.md

```diff
@@ -0,0 +1,500 @@
+# Stateless Ceph-CSI plugin implementation
+
+## Terminology
+
+- CO: Container orchestration system, communicates with CSI plugins using CSI
+  RPCs
+- NS: Node Service; CSI plugin service per node in the CO, responsible for
+  staging volumes for application pod consumption
+- CS: Controller Service; CSI plugin service responsible for creation,
+  deletion, snapshots, and expanding volumes for use in the CO system
+
+**NOTE:**
+
+- Discussion below may use k8s specific terminology, but should be extensible
+  to similar constructs on other container orchestration(CO) systems
+
+## Proposal/problem summary
+
+This document details a proposed solution, among presented alternatives, to
+implement a stateless Ceph-CSI plugin. It also details a naming scheme for
+volumes provisioned against a Ceph cluster and the resultant *volume_id* for
+the same that is shared with the CO system, enabling state to be carried by the
+*volume_id* where required.
+
+As of this writing, the plugin uses a config map to store *volume_name* to
+*volume_id* mapping, and also stores related Ceph cluster configuration
+information (e.g MON list, pool information) within the same tuple, to be able
+to find and operate on the provisioned volume. The implementation also depends
+on secrets from the StorageClass to perform NS related operations.
+
+The existing strategy enables a single CSI instance to operate across any Ceph
+cluster that is fully defined by the StorageClass, but brings in the burden of
+storing and managing provisioned volumes by the plugin.
+
+The proposal laid out here is to try and retain the ability for a single CSI
+instance to operate across any Ceph cluster, but remove the storage and
+management of provisioned CSI volumes by the plugin, or in short make the
+Ceph-CSI plugin stateless. Further, if this is achieved, it is also possible to
+run multiple replicas of the CS instance, to aid with load balancing of
+requests, than the current one instance limitation due to the existing config
+map.
+
+## Proposed solution details
+
+### Summary
+
+- Create a config map per Ceph cluster called `ceph-cluster-<cluster-fsid>` and
+  a pair of secrets to go along with the same cluster called
+  `ceph-cluster-<cluster-fsid>-provisioner-secret` and
+  `ceph-cluster-<cluster-fsid>-publish-secret`
+- Allow the CS and NS namespaces access to the above config maps and secrets,
```

Comment:
Are the CS and NS namespaces outside the scope of our own csi driver? I was hoping we could put everything rook needs for csi in the rook-ceph-system namespace along with the rook operator.